### DISCUSSION TITLE
Prism

### DISCUSSION THREAD

**Discussion by cptslow-lab** [mjhoptics/ray-optics#123] [Q&A]
Thanks a lot for the package. 
I failed to create a prism inside an optical model. Could you briefly outline how to do that.

Thanks

**Comment by mjhoptics** [comment]
Hello @ahochwal,

I apologize for the delay in responding, I was on the road. Here's an example with a right angle prism. The prism rendering in the lens picture is wrong, but the ray paths are (and should be) correct.

Sequential models of many prisms may be built using decenter data on the prism surfaces. Since the lens picture will be wrong, there are some list_* options in the **seq_model** such as list_sg() and list_gbl_tfrms() that list the decenter data with the system interfaces and gaps, and a listing of the global surface coords of the interface vertices.

The [notebook](https://github.com/mjhoptics/ray-optics-notebooks/blob/master/issues/issue123/issue123_prism.ipynb) is in my ray-optics-notebooks repo.

Feel free to ask questions. Thanks
Mike Hayford

# Investigate issue #123

## How do you construct a prism in ray-optics?

## Example 1: a right angle prism with a reflecting hypotenuse surface


```python
%matplotlib inline
# use widget to enable interactive figures.
#%matplotlib widget
```


```python
isdark = False
```


```python
from rayoptics.environment import *
```

## Create model from scratch


```python
opm = OpticalModel()
opm.radius_mode = True

sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']
```

## Optical Specs: f/5, 500mm efl, 1 deg fov, visual, monochromatic


```python
pupil_diameter = 100.
pupil_radius = pupil_diameter/2
osp.pupil = PupilSpec(osp, key=['object', 'pupil'], value=pupil_diameter)

osp.field_of_view = FieldSpec(osp, key=['object', 'angle'], value=1, 
                              is_relative=True, flds=[0.0, 1.0])

# wavelength for analysis: green, e-line
osp.spectral_region = WvlSpec([('e', 1.0)], ref_wl=0)
listobj(osp)
```

    'pupil' deprecated; use 'epd' instead.
    aperture: object epd; value=100.0
    field: object angle; value=1.0
    x,y=0.0 vlx= 0.000 vux= 0.000 vly= 0.000 vuy= 0.000
    y=1.0 vly= 0.000 vuy= 0.000 vlx= 0.000 vux= 0.000
    central wavelength=  546.0740 nm
    wavelength (weight) =  546.0740 (1.000)*
    focus shift=0.0
    


## Set up initial system - a singlet


```python
sm.do_apertures = False
sm.gaps[0].thi = 1e+11

efl = 500.
opm.add_lens(lbl='L1', power=1/efl, bending=.75, th=12.5, sd=105./2, med='N-BK7,Schott', t=500.)
sm.stop_surface = 1

opm.update_model()
```




```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+11       air             1      1.0000
     Stop:   294.780971      12.5000     N-BK7             1      52.500
        2: -2063.466800      500.000       air             1      52.500
      Img:     0.000000      0.00000                       1      1.0000



```python
sm.ifcs[-2].optical_power = pm.pwr_slope_solve(pm.ax, -2, -pupil_radius/efl)
opm.update_model(src_model=sm)

sm.gaps[-1].thi += pm.thi_ht_solve(pm.ax, -1, 0.)
opm.update_model(src_model=sm)
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+11       air             1      1.0000
     Stop:   294.780971      12.5000     N-BK7             1      52.500
        2: -2127.255675      492.758       air             1      52.500
      Img:     0.000000      0.00000                       1      1.0000



```python
pm.first_order_data()
```

    efl                 500
    ffl                -499
    pp1               1.003
    bfl               492.8
    ppk               7.242
    m                -5e-09
    red              -2e+08
    obj_dist          1e+11
    obj_ang               1
    enp_dist             -0
    enp_radius           50
    na obj            5e-10
    n obj                 1
    img_dist          492.8
    img_ht            8.728
    exp_dist         -8.247
    exp_radius         50.1
    f/# img               5
    na img          -0.0995
    n img                 1
    optical invariant       0.8728
    



```python
pm.list_model()
```

               ax_ht        pr_ht       ax_slp       pr_slp         power          tau        index    type
     0:            0 -1.74551e+09        5e-10    0.0174551             0        1e+11      1.00000    dummy
     1:           50            0   -0.0879843    0.0174551   0.001759686       8.2306      1.51872    transmit
     2:      49.2758     0.143666         -0.1      0.01742  0.0002438456      492.758      1.00000    transmit
     3: -7.10543e-15      8.72753         -0.1      0.01742             0            0      1.00000    dummy



```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
```


    
![output_15_0](https://github.com/mjhoptics/ray-optics/assets/33294031/efebde75-8fdc-4aed-b151-3dd0df876350)
    



```python
ta_plt = plt.figure(FigureClass=RayFanFigure, opt_model=opm, data_type='Ray', 
                     scale_type=Fit.User_Scale, user_scale_value=1., dpi=130, is_dark=isdark).plot()
```

    
![output_16_1](https://github.com/mjhoptics/ray-optics/assets/33294031/33ce1c05-bbd9-4b50-a4a6-fe078476b4dd)
    


## add plane-parallel plate


```python
sm.gaps[-1].thi = 300.
opm.add_lens(lbl='L1', power=0.0, th=75, sd=75./2, med='N-BK7,Schott', t=125.)

opm.update_model()

sm.gaps[-1].thi += pm.thi_ht_solve(pm.ax, -1, 0.)
opm.update_model(src_model=sm)
```


```python
sm.list_model()
```

                  r            t        medium     mode   zdr      sd
      Obj:     0.000000  1.00000e+11       air             1      1.0000
     Stop:   294.780971      12.5000     N-BK7             1      52.500
        2: -2127.255675      300.000       air             1      52.500
        3:     0.000000      75.0000     N-BK7             1      37.500
        4:     0.000000      143.375       air             1      37.500
      Img:     0.000000      0.00000                       1      1.0000



```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
```


    
![output_20_0](https://github.com/mjhoptics/ray-optics/assets/33294031/042d5e93-0b06-425d-8ddb-00a827559dde)
    



```python
ta_plt = plt.figure(FigureClass=RayFanFigure, opt_model=opm, data_type='Ray', 
                     scale_type=Fit.User_Scale, user_scale_value=1., dpi=130, is_dark=isdark).plot()
```



    
![output_21_1](https://github.com/mjhoptics/ray-optics/assets/33294031/5994f68d-f2cd-4108-aea4-d2bb5bfc3bdc)
    


## insert a reflecting surface as the prism hypotenuse


```python
sm.set_cur_surface(3)
sm.gaps[3].thi = 37.5
sm.add_surface([0, -37.5, 'refl'])
sm.ifcs[sm.cur_surface].decenter = srf.DecenterData('bend', alpha=45.)

opm.update_model()

sm.gaps[-1].thi += pm.thi_ht_solve(pm.ax, -1, 0.)
opm.update_model(src_model=sm)
```


```python
sm.list_sg()
```

                   r               mode              type          y       alpha
                           t           medium
      Obj:      0.00000                 
                     1.00000e+11          air
     Stop:      294.781                 
                         12.5000        N-BK7
        2:     -2127.26                 
                         300.000          air
        3:      0.00000                 
                         37.5000        N-BK7
        4:      0.00000          reflect               bend     0.0000     45.000
                        -37.5000        N-BK7
        5:     -0.00000                 
                        -143.375          air
      Img:     -0.00000                 



```python
sm.list_gbl_tfrms()
```

     0:    1.000000    0.000000    0.000000       0.00000
           0.000000    1.000000    0.000000       0.00000
           0.000000    0.000000    1.000000  -100000000000.00000
    
     1:    1.000000    0.000000    0.000000       0.00000
           0.000000    1.000000    0.000000       0.00000
           0.000000    0.000000    1.000000       0.00000
    
     2:    1.000000    0.000000    0.000000       0.00000
           0.000000    1.000000    0.000000       0.00000
           0.000000    0.000000    1.000000      12.50000
    
     3:    1.000000    0.000000    0.000000       0.00000
           0.000000    1.000000    0.000000       0.00000
           0.000000    0.000000    1.000000     312.50000
    
     4:    1.000000    0.000000    0.000000       0.00000
           0.000000    0.707107    0.707107       0.00000
           0.000000   -0.707107    0.707107     350.00000
    
     5:    1.000000    0.000000    0.000000       0.00000
           0.000000    0.000000    1.000000     -37.50000
           0.000000   -1.000000    0.000000     350.00000
    
     6:    1.000000    0.000000    0.000000       0.00000
           0.000000    0.000000    1.000000    -180.87473
           0.000000   -1.000000    0.000000     350.00000
    



```python
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=isdark).plot()
```


    
![output_26_0](https://github.com/mjhoptics/ray-optics/assets/33294031/6b49a3a0-3dbc-4e74-b1b4-08741be45c76)
    



```python
ta_plt = plt.figure(FigureClass=RayFanFigure, opt_model=opm, data_type='Ray', 
                     scale_type=Fit.User_Scale, user_scale_value=1., dpi=130, is_dark=isdark).plot()
```



    
![output_27_1](https://github.com/mjhoptics/ray-optics/assets/33294031/46fd7183-f3bb-4b04-a93d-6cf53816e43b)

**Reply by cptslow-lab** [reply]
Thank you for the example. Very helpful. Apparently, I was confused by the graphical representation in the first place. Is there a way to make the plot of the prism a bit more obvious? Maybe using dummy planes to indicate the actual position of the interface?

**Reply by mjhoptics** [reply]
Right now, there's no way to do this that I can think of that doesn't require a code change. The issue is the drawing of the prism itself; tilts/decenters of elements themselves are drawn correctly. 
The approach for this would be to add a new Part subclass to the [elements](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.elements.html#module-rayoptics.elem.elements) package. The range of surfaces comprising the prism would be inputs. A right angle prism would be straightforward to do, I think, but I don't think I can get to it for a few weeks, at the earliest.
Regards
Mike Hayford

**Reply by cptslow-lab** [reply]
Understood. Thank you for the information. I really appreciate it.
One final question. What does the bending keyword do in the add lens command?

**Reply by mjhoptics** [reply]
Bending is a dimensionless parameter describing the shape of a lens of a given power. 

B = (c1 + c2)/(c1 - c2)

- convex/concave, plano surface order, B = +1
- equi-convex/concave, B = 0
- plano, convex/concave surface order, B = -1
- meniscus bending right, B > +1
- meniscus bending left, B < -1

HTH, Mike
